src/url.cpp

```diff
@@ -0,0 +1,137 @@
+#include <algorithm>
+#include <string>
+#include <iterator>
+#include <unordered_map>
+#include <unordered_set>
+#include <iostream>
+
+#include "url.h"
+
+namespace Url
+{
+
+    /* Character classes */
+    const std::string Url::GEN_DELIMS = ":/?#[]@";
+    const std::string Url::SUB_DELIMS = "!$&'()*+,;=";
+    const std::string Url::ALPHA = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz";
+    const std::string Url::DIGIT = "0123456789";
+    const std::string Url::UNRESERVED = Url::ALPHA + Url::DIGIT + "-._~";
+    const std::string Url::RESERVED = Url::GEN_DELIMS + Url::SUB_DELIMS;
+    const std::string Url::PCHAR = Url::UNRESERVED + Url::SUB_DELIMS + ":@";
+    const std::string Url::PATH = Url::PCHAR + "/";
+    const std::string Url::QUERY = Url::PCHAR + "/?";
+    const std::string Url::FRAGMENT = Url::PCHAR + "/?";
+    const std::string Url::USERINFO = Url::UNRESERVED + Url::SUB_DELIMS + ":";
+    const std::string Url::SCHEME =
+        "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789+-.";
+
+    Url::Url(const std::string& url): port_(0)
+    {
+        size_t position = 0;
+        size_t index = url.find(':');
+        if (index != std::string::npos)
+        {
+            // All the characters in our would-be scheme must be in SCHEME
+            if (url.substr(0, index).find_first_not_of(SCHEME) == std::string::npos)
+            {
+                // If there is nothing after the : or there are any non-digits, this is
+                // the scheme
+                if ((index + 1) >= url.length()
+                    || url.find_first_not_of(DIGIT, index + 1) != std::string::npos)
+                {
+                    scheme_ = url.substr(0, index);
+                    std::transform(
+                        scheme_.begin(), scheme_.end(), scheme_.begin(), ::tolower);
+                    position = index + 1;
+                }
+            }
+        }
+
+        // Search for the netloc
+        if (url.substr(position, 2).compare("//") == 0)
+        {
+            // Skip the '//'
+            position += 2;
+            index = url.find_first_of("/?#", position);
+            host_ = url.substr(position, index - position);
```

Comment:
That's if `pos` is beyond `size()`. Otherwise:

> Returns a substring [pos, pos+count). If the requested substring extends past the end of the string, or if count == npos, the returned substring is [pos, size()).